File: Snooper/src/main/java/com/prateekj/snooper/customviews/PaginatedRecyclerView.java
package com.prateekj.snooper.customviews;


import android.content.Context;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.AttributeSet;

public class PaginatedRecyclerView extends RecyclerView implements PageAddedListener {

  private NextPageRequestListener listener;
  private boolean loading;

  public PaginatedRecyclerView(Context context) {
    super(context);
  }

  public PaginatedRecyclerView(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
  }

  public PaginatedRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
    super(context, attrs, defStyle);
  }

  @Override
  public void setAdapter(Adapter adapter) {
    super.setAdapter(adapter);
    adapter.registerAdapterDataObserver(new AdapterDataAppendObserver(this));
  }

  public void setNextPageListener(NextPageRequestListener listener) {
    this.listener = listener;
  }

  @Override
  public void onScrolled(int dx, int dy) {
    super.onScrolled(dx, dy);
    if (loading || listener == null) {
      return;
    }
    if (needToRequestNextPage() && !listener.areAllPagesLoaded()) {
      loading = true;
      listener.requestNextPage();
    }
  }

  private boolean needToRequestNextPage() {
    LinearLayoutManager layoutManager = (LinearLayoutManager) getLayoutManager();
    int visibleItemCount = layoutManager.getChildCount();
    int totalItemCount = getAdapter().getItemCount();
    int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();
    return firstVisibleItemPosition > 0 && firstVisibleItemPosition + visibleItemCount > totalItemCount - 5;
  }

  @Override
  public void onPageAdded() {
    loading = false;
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/customviews/PaginatedRecyclerView.kt
package com.prateekj.snooper.customviews


import android.content.Context
import android.util.AttributeSet
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class PaginatedRecyclerView : RecyclerView, PageAddedListener {

  private var listener: NextPageRequestListener? = null
  private var loading: Boolean = false

  constructor(context: Context) : super(context)

  constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

  constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(
    context,
    attrs,
    defStyle
  )

  override fun setAdapter(adapter: Adapter<*>?) {
    super.setAdapter(adapter)
    adapter!!.registerAdapterDataObserver(AdapterDataAppendObserver(this))
  }

  fun setNextPageListener(listener: NextPageRequestListener) {
    this.listener = listener
  }

  override fun onScrolled(dx: Int, dy: Int) {
    super.onScrolled(dx, dy)
    if (loading || listener == null) {
      return
    }
    if (needToRequestNextPage() && !listener!!.areAllPagesLoaded()) {
      loading = true
      listener!!.requestNextPage()
    }
  }

  private fun needToRequestNextPage(): Boolean {
    val layoutManager = layoutManager as LinearLayoutManager?
    val visibleItemCount = layoutManager!!.childCount
    val totalItemCount = adapter!!.itemCount
    val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()
    return firstVisibleItemPosition > 0 && firstVisibleItemPosition + visibleItemCount > totalItemCount - 5
  }

  override fun onPageAdded() {
    loading = false
  }
}


